database/triggers.py

```diff
@@ -0,0 +1,213 @@
+        trgfunc_definition = f"""
+        RETURNS TRIGGER AS $$
+        BEGIN
+            INSERT INTO hame.lifecycle_date
+                (lifecycle_status_id, {table}_id, starting_at)
+            VALUES (NEW.lifecycle_status_id, NEW.id, CURRENT_TIMESTAMP);
+            RETURN NEW;
+        END;
+        $$ language 'plpgsql'
+        """
+
+        trg_signature = f"trg_{table}_new_lifecycle_date"
+        trg_definition = f"""
+        BEFORE UPDATE ON {table}
+        FOR EACH ROW
+        WHEN (NEW.lifecycle_status_id <> OLD.lifecycle_status_id)
+        EXECUTE FUNCTION hame.{trgfunc_signature}
+        """
+
+        trgfunc = PGFunction(
+            schema="hame",
+            signature=trgfunc_signature,
+            definition=trgfunc_definition,
+        )
+        new_lifecycle_date_trgfuncs.append(trgfunc)
+
+        trg = PGTrigger(
+            schema="hame",
+            signature=trg_signature,
+            on_entity=f"hame.{table}",
+            is_constraint=False,
+            definition=trg_definition,
+        )
+        new_lifecycle_date_trgs.append(trg)
+
+    return new_lifecycle_date_trgs, new_lifecycle_date_trgfuncs
+
+
+def generate_update_lifecycle_status_triggers():
+    update_lifecycle_status_trgs = []
+    update_lifecycle_status_trgfuncs = []
+
+    for table in tables_with_lifecycle_status:
+        trgfunc_signature = f"trgfunc_{table}_update_lifecycle_status()"
+        trgfunc_definition = f"""
+        RETURNS TRIGGER AS $$
+        BEGIN
+            UPDATE hame.{table}
+            SET lifecycle_status_id = NEW.lifecycle_status_id
+            WHERE plan_regulation_group_id = NEW.plan_regulation_group_id;
+            RETURN NEW;
+        END;
+        $$ language 'plpgsql'
+        """
+
+        trg_signature = f"trg_{table}_update_lifecycle_status"
+        trg_definition = f"""
+        BEFORE UPDATE ON plan
+        FOR EACH ROW
+        WHEN (NEW.lifecycle_status_id <> OLD.lifecycle_status_id)
+        EXECUTE FUNCTION hame.{trgfunc_signature}
+        """
+
+        trgfunc = PGFunction(
+            schema="hame", signature=trgfunc_signature, definition=trgfunc_definition
+        )
+        update_lifecycle_status_trgfuncs.append(trgfunc)
+
+        trg = PGTrigger(
+            schema="hame",
+            signature=trg_signature,
+            on_entity="hame.plan",
+            is_constraint=False,
+            definition=trg_definition,
+        )
+        update_lifecycle_status_trgs.append(trg)
+
+    return update_lifecycle_status_trgs, update_lifecycle_status_trgfuncs
+
+
+def generate_add_plan_id_fkey_triggers():
+    add_plan_id_fkey_trgs = []
+    add_plan_id_fkey_trgfuncs = []
+
+    for table in plan_object_tables:
+        trgfunc_signature = f"trgfunc_{table}_add_plan_id_fkey()"
+        trgfunc_definition = """
+        RETURNS TRIGGER AS $$
+        BEGIN
+            -- Get the row with most recent created_at timestamp in plan table
```

Comment:
Let's do this so that instead of selecting the latest plan, we will select the plan that contains the object geometry.

I think this can be done using the https://postgis.net/docs/ST_Contains.html function. So let's select the latest plan that contains the geometry.

If no plans are found that contain the object, we should actually prevent saving the object, because every object should be inside a plan. So I guess we can use is_constraint=True for this trigger.